File: trigger.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <unistd.h>
#include <inttypes.h>
#include <poll.h>

#include "trigger.h"

void zz_trigger_write(zz_trigger *t, uint64_t data) {
  int nbytes = write(t->fd, &data, sizeof(uint64_t));
  if (nbytes != 8) {
    fprintf(stderr, "zz_trigger_fire(): cannot write to event fd\n");
    exit(1);
  }
}

void zz_trigger_fire(zz_trigger *t) {
  zz_trigger_write(t, 1);
}

void zz_trigger_poll(zz_trigger *t) {
  struct pollfd fds[1];
  fds[0].fd = t->fd;
  fds[0].events = POLLIN;
  int status = poll(fds, 1, -1);
  if (status != 1) {
    fprintf(stderr, "zz_trigger_poll(): status=%d, expected 1\n", status);
    exit(1);
  }
}

uint64_t zz_trigger_read(zz_trigger *t) {
  uint64_t data = 0;
  int nbytes = read(t->fd, &data, 8);
  if (nbytes != 8) {
    fprintf(stderr, "zz_trigger_read(): nbytes=%d, expected 8\n", nbytes);
    exit(1);
  }
  return data;
}

uint64_t zz_trigger_wait(zz_trigger *t) {
  zz_trigger_poll(t);
  return zz_trigger_read(t);
}

File: trigger.lua
local ffi = require('ffi')
local sched = require('sched')
local util = require('util')
local errno = require('errno')

ffi.cdef [[

enum {
  EFD_SEMAPHORE = 00000001,
  EFD_CLOEXEC   = 02000000,
  EFD_NONBLOCK  = 00004000
};

int eventfd(unsigned int initval, int flags);

typedef struct {
  int fd;
} zz_trigger;

]]

local M = {}

local Trigger_mt = {}

function Trigger_mt:read()
   local buf = ffi.new("uint64_t[1]", 0)
   local nbytes = ffi.C.read(self.fd, buf, 8)
   if nbytes == 8 then
      local value = tonumber(buf[0])
      assert(value > 0)
      return value
   elseif nbytes == -1 then
      local errnum = errno.errno()
      if errnum ~= ffi.C.EAGAIN then
         ef("read() failed: %s", errno.strerror(errnum))
      end
      return nil
   else
      ef("read() failed: nbytes=%d, expected 8", nbytes)
   end
end

function Trigger_mt:poll()
   local rv = nil
   while not rv do
      if sched.ticking() then
         sched.poll(self.fd, "r")
      end
      rv = self:read()
   end
   return rv
end

function Trigger_mt:write(data)
   local buf = ffi.new("uint64_t[1]")
   buf[0] = data
   ffi.C.write(self.fd, buf, 8)
end

function Trigger_mt:fire()
   self:write(1)
end

function Trigger_mt:delete()
   if self.fd ~= 0 then
      ffi.C.close(self.fd)
      self.fd = 0
   end
end

Trigger_mt.__index = Trigger_mt
Trigger_mt.__gc = Trigger_mt.delete

local Trigger = ffi.metatype("zz_trigger", Trigger_mt)

function M.Trigger()
   local fd = util.check_errno("eventfd",
      ffi.C.eventfd(0, ffi.C.EFD_NONBLOCK))
   return Trigger(fd)
end

function M.Semaphore()
   local fd = util.check_errno("eventfd",
      ffi.C.eventfd(0, bit.bor(ffi.C.EFD_NONBLOCK, ffi.C.EFD_SEMAPHORE)))
   return Trigger(fd)
end

return setmetatable(M, { __call = M.Trigger })
